### PR TITLE
tests: Add sink-kafka-restart test

### PR DESCRIPTION
**Pull request by def-**
Maybe I'm missing something. This currently works with --redpanda. I followed the description from https://materializeinc.slack.com/archives/C01LKF361MZ/p1720682554959769

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by def-** [comment]
Thanks Nikhil! Verified locally that redpanda indeed fails the test now. Triggered a full nightly because of the kafka/redpanda upgrades too: https://buildkite.com/materialize/nightly/builds/8582

File: src/storage/src/sink/kafka.rs [hunk 1]
```diff
@@ -409,6 +409,8 @@ impl TransactionalProducer {
             .key(&self.progress_key);
         self.producer.send(record).map_err(|(e, _)| e)?;
 
+        fail::fail_point!("kafka_sink_commit_transaction");
```
Comment (nrainer-materialize) [inline src/storage/src/sink/kafka.rs:412]: Is this change intended?

Comment (nrainer-materialize) [reply src/storage/src/sink/kafka.rs:412]: Ah, this is the change by Nikhil.

**Review by nrainer-materialize — APPROVED** [review]

File: src/storage/src/sink/kafka.rs (hunk 1 above)
Comment (def-) [reply src/storage/src/sink/kafka.rs:412]: Yes! It adds the fail point, where we sleep for 5 s in the test, which is the way to make the test reliable.
